pilot/pkg/dns/dns.go

```diff
@@ -354,8 +355,23 @@ func generateAltHosts(hostname string, nameinfo *nds.NameTable_NameInfo, proxyNa
 // of registry, we will look it up in one of our tables, failing which we will return NXDOMAIN.
 func (table *LookupTable) lookupHost(qtype uint16, hostname string) ([]dns.RR, bool) {
 	var hostFound bool
-	if _, hostFound = table.allHosts[hostname]; !hostFound {
-		// this is not from our registry
+
+	question := host.Name(hostname)
+	wildcard := false
+	// First check if host exists.
+	_, hostFound = table.allHosts[hostname]
+	if !hostFound {
+		// Check matching hosts for wildcard hosts.
+		for h := range table.allHosts {
```

Comment:
O(N) lookup on the request path seems a bit concerning. Doesn't DNS only allow a single label? So we can store `*.example.com` as `table.wildcardHosts["example.com"]` and strip off the first label and do an O(1) lookup? Or better, whatever coreDNS does :slightly_smiling_face: 

Not sure if `*.*.example.com` is valid though...

Reply:
I will take a look at coreDNS and change it accordingly :-)

Reply:
Looked at core DNS - core DNS only replaces the first label - which is easy to handle i.e. when request comes for "foo.example.com" we can lookup "*.example.com" and match.

However in our case it could be "svc.svcns.domainname" and our lookup table will have "*.domainname" - I do not know it is ok to replace two labels with * and match with *.domainname?

@hzxuzhonghu Any thoughts?

Reply:
> However in our case it could be "svc.svcns.domainname" and our lookup table will have "*.domainname" - I do not know it is ok to replace two labels with * and match with *.domainname?


I don't think the DNS spec permits this. I could be wrong though

Reply:
I stand corrected:
```
another way
to look at this is that the "*" label always matches at least one whole
label and sometimes more, but always whole labels.
```
https://datatracker.ietf.org/doc/html/rfc1034#section-4.3.3

I confirmed GCP cloud DNS supports it as well

Reply:
If that is the case, we need linear search?

Reply:
The search could be O(labels) on the request (for a.b.c.d, lookup a.b.c.d, *.b.c.d, *.c.d, etc) instead of O(hosts). Or at least O(wildcard hosts) (keep a map of wildcard hosts).

Just plain O(hosts) seems pretty bad since there are a lot of hosts

Reply:
But I think we can also do O(labels)